prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/HttpResponseCallback.java

```diff
@@ -0,0 +1,192 @@
+import static org.apache.flink.connector.prometheus.sink.SinkMetrics.SinkCounter.NUM_SAMPLES_OUT;
+import static org.apache.flink.connector.prometheus.sink.SinkMetrics.SinkCounter.NUM_SAMPLES_RETRY_LIMIT_DROPPED;
+import static org.apache.flink.connector.prometheus.sink.SinkMetrics.SinkCounter.NUM_WRITE_REQUESTS_OUT;
+import static org.apache.flink.connector.prometheus.sink.SinkMetrics.SinkCounter.NUM_WRITE_REQUESTS_PERMANENTLY_FAILED;
+
+/**
+ * Callback handling the outcome of the http async request.
+ *
+ * <p>This class implements the error handling behaviour, based on the configuration in {@link
+ * SinkWriterErrorHandlingBehaviorConfiguration}. Depending on the condition, the sink may throw an
+ * exception and cause the job to fail, or log the condition to WARN, increment the counters and
+ * continue with the next request.
+ *
+ * <p>In any case, every write-request either entirely succeed or fail. Partial failures are not
+ * handled.
+ *
+ * <p>In no condition a write-request is re-queued for the AsyncSink to reprocess: this would cause
+ * out of order writes that would be rejected by Prometheus.
+ *
+ * <p>Note that the http async client retries, based on the configured retry policy. The callback is
+ * called with an outcome of *completed* either when the request has succeeded or the max retry
+ * limit has been exceeded. It is responsibility of the callback distinguishing between these
+ * conditions.
+ */
+class HttpResponseCallback implements FutureCallback<SimpleHttpResponse> {
+    private static final Logger LOG = LoggerFactory.getLogger(HttpResponseCallback.class);
+
+    private final int timeSeriesCount;
+    private final long sampleCount;
+    private final Consumer<List<Types.TimeSeries>> reQueuedResult;
+    private final SinkMetrics metrics;
+    private final SinkWriterErrorHandlingBehaviorConfiguration errorHandlingBehaviorConfig;
+
+    public HttpResponseCallback(
+            int timeSeriesCount,
+            long sampleCount,
+            SinkMetrics metrics,
+            SinkWriterErrorHandlingBehaviorConfiguration errorHandlingBehaviorConfig,
+            Consumer<List<Types.TimeSeries>> reQueuedResult) {
+        this.timeSeriesCount = timeSeriesCount;
+        this.sampleCount = sampleCount;
+        this.reQueuedResult = reQueuedResult;
+        this.metrics = metrics;
+        this.errorHandlingBehaviorConfig = errorHandlingBehaviorConfig;
+    }
+
+    /**
+     * The completed outcome is invoked every time the http client successfully receives a valid
+     * http response, regardless of the status code.
+     *
+     * <p>This method classifies the responses and implements the behaviour expected by the
+     * Remote-Write specifications. In case of error, the behaviour is determined by the error
+     * handling configuration.
+     */
+    @Override
+    public void completed(SimpleHttpResponse response) {
```

Comment:
Do we want to consider expanding out nested `if-else` into linear with `return` instead?

See example here: https://stackoverflow.com/questions/56356190/nested-if-conditions-vs-multiple-separated-if-conditions-with-return-statement-i

Reply:
Done. 
One nested if (no else) is still required to throw exception when the error behaviour is FAIL, avoiding too many `if (A && X1)... if (A && X2)...`